### DISCUSSION TITLE
Jupiter hiccup

### DISCUSSION THREAD

**Discussion by cderi** [skyfielders/python-skyfield#815] [Q&A] [answered]
Many thanks for this great library that I discovered recently.

While doing some tests I found this strange result on Jupiter trajectory around 2032-01-01 07:40.
This is the plot of ra,radians  from  earth.at(t).observe(jupiter).apparent().radec('date') using de421.bsp.
I could not figure-out what could be the reason. I hope this is not too obvious to ask here !
Many Thanks.

![Jupiter 2032-01-01 07h40](https://user-images.githubusercontent.com/52795294/205400564-73033e1c-5c29-4d46-b7bc-f2b75b90e573.png)

**Comment by Bernmeister** [comment]
Whilst I have no idea how/what to suggest in solving, it would greatly benefit others who would if you could please post the entire script(s) which generated the results and produced the graph?  If you paste in the code and then place three [\`\`\`](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) above and below, that'll format things nicely for you (you can then also preview to see how it turns out).

**Comment by brandon-rhodes** [comment, marked as answer]
Looking at the situation with Stellarium, it looks like Jupiter goes behind the Sun on that date. My guess would be that the Sun's gravity deflects Jupiter's image enough to cause the small wiggle in your graph.

**Reply by cderi** [reply]
Great catch, many thanks !

The wiggle is 1 minute angle so I find it quite big. Shouldn't the Sun's gravity deflexion be in a few seconds range ?

For anyone interested in looking at it, here is the script. I find it funny that it happens on Jan. 1st !

```
from skyfield.api import load
import matplotlib.pyplot as plt

ts = load.timescale()
planets = load('de421.bsp')

earth  = planets['earth']
jupiter = planets['jupiter barycenter']

t0 = ts.ut1(2032, 1, 1, 5, 0, 0)
t1 = ts.ut1(2032, 1, 1, 10, 0, 0)
t = ts.linspace(t0, t1, 301)

ra =  earth.at(t).observe(jupiter).apparent().radec('date')[0]
plt.plot(t.ut1, ra.radians, '-')
plt.show()
```

**Reply by cderi** [reply]
Note that the IMCCE calculation does not show the wiggle ([Link](https://ssp.imcce.fr/forms/observing)) with the same parameters. It is supposed to take into account the deviation from the Sun.

Lieu d'observation : 500 (Code observatoire)
Époque : 2032-01-01T06:00:00.000 / +301 / 1 Minute / UTC
Corps du système solaire : Jupiter (Planet)
Éphémérides de position

**Reply by davidmikolas** [reply]
The wiggle doesn't show in Horizons, note that the Sun-Observer-Target (SOT) angle goes down to 0.077 degrees, which puts the line of site passing straight through the Sun, passing within ~200,000 km of the center (the Sun's radius i s ~700,000 km) I guess the deflection calculation here puts the mass of the Sun at a single point. 

I wonder if there are flags possible indicating when a deflection calculation is wrong because it's this unphysical?

<img width="497" alt="Screen Shot 2022-12-04 at 8 24 05 AM" src="https://user-images.githubusercontent.com/17236804/205467909-3c0b0c33-7658-481f-a710-734e8918a98e.png">

**Reply by brandon-rhodes** [reply]
Here's an open issue you can follow about Skyfield's overly exuberant deflection once a body is behind the disc of a mass like the Sun or Jupiter:

https://github.com/skyfielders/python-skyfield/issues/734

As Skyfield uses the same geometry logic as the USNO "NOVAS" library, I expect that NOVAS has the same problem. I wonder if we should take a look at how, say, the SPICE source code handles deflection once the bodies are very close?

**Reply by davidmikolas** [reply]
This is _interesting!_ 

I think that 1) continuous position information should be returned to avoid solver hiccups, 2) an _at-least-better_ and _neutrino-friendly_ model for trajectories near and through the Sun is out there and should be used if possible, and 3) it would be great/cool if there was a mechanism for letting users know when the light path grazes or passes through a massive object.

I'm going to a) poke around to see what I can understand about the physics of the deflection and the implementation in "USNO’s NOVAS Fortran code" and b) ask an Astronomy SE question where I know there are some GR-savvy astronomers + space scientists who may find this interesting. Gravitational lensing deflection to first order is just 1/r and that can be applied to mass distributions (see arXiV papers linked in https://astronomy.stackexchange.com/questions/50029/if-i-cant-unscramble-an-egg-how-do-astronomers-unscramble-views-gravitationall) and I guess since Newton's shell theorem applies only to 1/r^2 forces the integral has to always be over the whole sphere, not just the mass inside r.

**Reply by davidmikolas** [reply]
posted: https://astronomy.stackexchange.com/questions/51200/light-or-neutrinos-graze-or-pass-through-the-sun-and-arrive-at-earth-need-an-e

**Comment by Bernmeister** [comment]
@brandon-rhodes Didn't even need a script; damn you're good!
